casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala

```diff
@@ -0,0 +1,131 @@
+  def addApproval(a: BlockApproval): F[Unit] = {
+    val validSig = for {
+      c   <- a.candidate if c == this.candidate
+      sig <- a.sig if Validate.signature(sigData, sig)
+    } yield sig
+    val sender = a.sig.fold("<Empty Signature>")(sig => Base16.encode(sig.publicKey.toByteArray))
+
+    validSig.fold(Log[F].warn(s"APPROVAL: ignoring invalid block approval from $sender")) { sig =>
+      sigsF.update(_ + sig) *> Log[F].info(s"APPROVAL: received block approval from $sender")
+    }
+  }
+
+  def currentSigs: F[Set[Signature]] = sigsF.get
+
+  def approvedBlock: F[Option[ApprovedBlock]] =
+    sigsF.get.map(
+      sigs =>
+        if (sigs.size >= requiredSigs) Some(ApprovedBlock(Some(candidate), sigs.toSeq))
+        else none[ApprovedBlock])
+
+  //TODO: potential optimization, only send to peers we have not
+  //      reveived a valid signature from yet
+  def sendUnapprovedBlock: F[Unit] =
+    for {
+      _ <- Log[F].info(s"APPROVAL: Beginning send of UnapprovedBlock $candidateHash to peers...")
+      _ <- CommUtil.sendToPeers[F](transport.UnapprovedBlock, serializedUnapprovedBlock)
+      _ <- Log[F].info(s"APPROVAL: Sent UnapprovedBlock $candidateHash to peers.")
+    } yield ()
+
+  def sendApprovedBlock: F[Boolean] =
+    approvedBlock.flatMap {
+      case Some(a) =>
+        val serializedApprovedBlock = a.toByteString
+        for {
+          _ <- Log[F].info(s"APPROVAL: Beginning send of ApprovedBlock $candidateHash to peers...")
+          _ <- CommUtil.sendToPeers[F](transport.ApprovedBlock, serializedApprovedBlock)
+          _ <- Log[F].info(s"APPROVAL: Sent ApprovedBlock $candidateHash to peers.")
+        } yield true
+
+      case None => false.pure[F]
+    }
+
+}
+
+object ApproveBlockProtocol {
+
+  def apply[F[_]](implicit instance: ApproveBlockProtocol[F]): ApproveBlockProtocol[F] = instance
+
+  def create[F[_]: Capture: Sync: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler](
+      block: BlockMessage,
+      requiredSigs: Int,
+      duration: Long): F[ApproveBlockProtocol[F]] =
+    for {
+      now   <- Time[F].currentMillis
+      sigsF <- Ref.of[F, Set[Signature]](Set.empty[Signature])
+    } yield new ApproveBlockProtocol(block, requiredSigs, now, duration, sigsF)
+
+  def blockApprovalPacketHandler[F[_]: Monad: ApproveBlockProtocol](
+      peer: PeerNode): PartialFunction[Packet, F[Option[Packet]]] =
+    Function
+      .unlift(packetToBlockApproval)
+      .andThen {
+        case ba: BlockApproval =>
+          for {
+            _ <- ApproveBlockProtocol[F].addApproval(ba)
+          } yield none[Packet]
+      }
+
+  def run[F[_]: Monad: Capture: Time: ApproveBlockProtocol](waitTime: Long = 5000L): F[Unit] =
+    for {
+      t             <- Time[F].currentMillis
+      maybeApproved <- ApproveBlockProtocol[F].approvedBlock
+      _ <- maybeApproved match {
+            case Some(_)
+                if (ApproveBlockProtocol[F].candidate.requiredSigs == 0 || t >= ApproveBlockProtocol[
```

Comment:
Is `ApproveBlockProtocol[F].candidate.requiredSigs == 0` some special case?

Reply:
Yes. I had the same question originally  too. Probably means we should add a comment saying how it is unnecessary to wait for sigs if we don't require any in the first place.

Reply:
@KentShikama What would be an example of the block that requires no signatures?

Reply:
It's for testing. The `requiredSigs == 0` case will function identically to how the node behaves now where it just gets to create the approved block immediately and send it out. In production `requiredSigs` will certainly be positive.

Reply:
The other thing is - I thought that this protocol is only for genesis block?

Reply:
The other thing is - I thought that this protocol is only for genesis block?

Reply:
It is also for checkpoints along the way under assumption of weak subjectivity. See https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity/